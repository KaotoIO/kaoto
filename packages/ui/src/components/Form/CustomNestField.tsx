import { Card, CardBody } from '@patternfly/react-core/dist/js/components/Card';
import { ConnectedFieldProps, connectField, filterDOMProps, HTMLFieldProps } from 'uniforms';
import { CustomAutoField } from './CustomAutoField';

type CustomNestFieldProps = HTMLFieldProps<object, HTMLDivElement, { helperText?: string; itemProps?: object }>;

export const CustomNestField = connectField(
  ({
    children,
    error,
    errorMessage,
    fields,
    itemProps,
    label,
    name,
    showInlineError,
    disabled,
    ...props
  }: ConnectedFieldProps<CustomNestFieldProps>) => {
    return (
      <Card {...filterDOMProps(props)}>
        <CardBody className="pf-c-form">
          {label && (
            <label>
              <b>{label}</b>
            </label>
          )}
          {children ||
            fields?.map((field) => <CustomAutoField key={field} disabled={disabled} name={field} {...itemProps} />)}
        </CardBody>
      </Card>
    );
  },
);
