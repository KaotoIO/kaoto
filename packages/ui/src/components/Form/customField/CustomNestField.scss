.custom-nest-field {
  border-radius: 15px;

  /* stylelint-disable-next-line selector-class-pattern */
  .pf-v5-c-card__body {
    display: grid;
    gap: var(--pf-v5-c-form--GridGap);
  }

  .pf-v5-c-expandable-section {
    --pf-v5-c-expandable-section__content--MarginTop: 0;
  }
}
