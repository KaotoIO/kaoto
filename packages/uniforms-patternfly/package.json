{
  "name": "@kaoto-next/uniforms-patternfly",
  "version": "0.0.0",
  "description": "Patternfly forms for uniforms",
  "license": "Apache-2.0",
  "author": {
    "name": "Gianluca",
    "email": "gzuccare@redhat.com"
  },
  "keywords": [],
  "homepage": "https://github.com/KaotoIO/kaoto-next/tree/main/packages/uniforms-patternfly",
  "repository": {
    "type": "git",
    "url": "https://github.com/KaotoIO/kaoto-next.git"
  },
  "bugs": {
    "url": "https://github.com/KaotoIO/kaoto-next/issues"
  },
  "types": "./dist/esm/index.d.ts",
  "main": "src/index.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rimraf dist && tsc --build tsconfig.cjs.json && tsc --build tsconfig.esm.json",
    "test": "jest"
  },
  "dependencies": {
    "@types/invariant": "^2.2.35",
    "invariant": "^2.2.4",
    "lodash": "^4.17.21",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "uniforms": "4.0.0-alpha.0"
  },
  "devDependencies": {
    "@babel/core": "^7.21.8",
    "@babel/preset-env": "^7.21.5",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.21.5",
    "@patternfly/react-core": "^5.0.0",
    "@patternfly/react-icons": "^5.0.0",
    "@testing-library/dom": "^9.3.0",
    "@testing-library/jest-dom": "^6.0.0",
    "@testing-library/react": "^14.0.0",
    "@testing-library/user-event": "^14.4.3",
    "@types/jest": "^29.4.0",
    "@types/lodash": "^4.14.168",
    "@types/node": "^18.0.0",
    "@types/react": "^17.0.6",
    "@types/react-dom": "^17.0.5",
    "@types/simpl-schema": "^1.12.0",
    "@types/testing-library__jest-dom": "^5.9.1",
    "@types/testing-library__react": "^10.0.0",
    "babel-jest": "^29.4.2",
    "copy-webpack-plugin": "^11.0.0",
    "jest": "^29.4.2",
    "jest-environment-jsdom": "^29.4.2",
    "rimraf": "^3.0.2",
    "simpl-schema": "^3.4.1",
    "ts-jest": "^29.1.1",
    "ts-node": "^10.9.1",
    "typescript": "^5.0.2",
    "uniforms-bridge-simple-schema-2": "4.0.0-alpha.1",
    "webpack": "^5.88.2",
    "webpack-cli": "^4.10.0",
    "webpack-dev-server": "^4.15.1",
    "webpack-merge": "^5.9.0"
  }
}
