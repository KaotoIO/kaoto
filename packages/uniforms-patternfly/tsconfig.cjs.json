{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "baseUrl": "src",
    "outDir": "dist/cjs",
    "rootDir": "src",
    "module": "commonjs",
    "target": "es6",
    "esModuleInterop": true,
    "moduleResolution": "node"
  },
  "include": ["src"]
}
