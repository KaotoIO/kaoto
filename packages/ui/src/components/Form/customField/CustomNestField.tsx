/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Card, CardBody, CardHeader, CardTitle } from '@patternfly/react-core';
import { useContext } from 'react';
import { HTMLFieldProps, connectField, filterDOMProps } from 'uniforms';
import { getFieldGroups } from '../../../utils';
import { CustomAutoField } from '../CustomAutoField';
import './CustomNestField.scss';
import { FilteredFieldContext } from '../../../providers';
import { CustomExpandableSection } from './CustomExpandableSection';

export type CustomNestFieldProps = HTMLFieldProps<
  object,
  HTMLDivElement,
  { properties?: Record<string, unknown>; helperText?: string; itemProps?: object }
>;

export const CustomNestField = connectField(
  ({
    children,
    error,
    errorMessage,
    fields,
    itemProps,
    label,
    name,
    showInlineError,
    disabled,
    ...props
  }: CustomNestFieldProps) => {
    const { filteredFieldText, isGroupExpanded } = useContext(FilteredFieldContext);
    const cleanQueryTerm = filteredFieldText.replace(/\s/g, '').toLowerCase();
    const filteredProperties = Object.entries(props.properties ?? {}).filter((field) =>
      field[0].toLowerCase().includes(cleanQueryTerm),
    );
    const actualProperties = Object.fromEntries(filteredProperties);
    const propertiesArray = getFieldGroups(actualProperties);

    if (propertiesArray.common.length === 0 && Object.keys(propertiesArray.groups).length === 0) return null;

    return (
      <Card isFlat className="custom-nest-field" data-testid={'nest-field'} {...filterDOMProps(props)}>
        <CardHeader>
          <CardTitle>{label}</CardTitle>
        </CardHeader>
        <CardBody>
          {propertiesArray.common.map((field) => (
            <CustomAutoField key={field} name={field} />
          ))}
        </CardBody>

        {Object.entries(propertiesArray.groups)
          .sort((a, b) => (a[0] === 'advanced' ? 1 : b[0] === 'advanced' ? -1 : 0))
          .map(([groupName, groupFields]) => (
            <CustomExpandableSection
              key={`${groupName}-section-toggle`}
              groupName={groupName}
              isGroupExpanded={isGroupExpanded}
            >
              <CardBody>
                {groupFields.map((field) => (
                  <CustomAutoField key={field} name={field} />
                ))}
              </CardBody>
            </CustomExpandableSection>
          ))}
      </Card>
    );
  },
);
