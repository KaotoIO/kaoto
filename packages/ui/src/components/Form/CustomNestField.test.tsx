import { render } from '@testing-library/react';
import { CustomNestField } from './CustomNestField';
import { AutoForm } from '@kaoto-next/uniforms-patternfly';
import { SchemaService } from '.';

describe('CustomNestField', () => {
  it('should render', () => {
    const schema = new SchemaService().getSchemaBridge({
      title: 'Rest Configuration',
      description: 'To configure rest',
      type: 'object',
      additionalProperties: false,
      properties: {
        objectProp: {
          type: 'object',
        },
      },
    });

    const { container } = render(
      <AutoForm schema={schema} onSubmit={() => {}} model={{}}>
        <CustomNestField name="objectProp" label="Custom nest field label" />
      </AutoForm>,
    );

    expect(container).toMatchSnapshot();
  });
});
