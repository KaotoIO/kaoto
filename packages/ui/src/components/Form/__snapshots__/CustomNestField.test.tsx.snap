// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomNestField should render 1`] = `
<div>
  <form
    class="pf-v5-c-form"
    data-testid="base-form"
    novalidate=""
  >
    <div
      class="pf-v5-c-card"
      data-ouia-component-id="OUIA-Generated-Card-1"
      data-ouia-component-type="PF5/Card"
      data-ouia-safe="true"
      id="uniforms-0000-0000"
    >
      <div
        class="pf-v5-c-card__body pf-c-form"
      >
        <label>
          <b>
            Custom nest field label
          </b>
        </label>
      </div>
    </div>
  </form>
</div>
`;
